velox/functions/prestosql/aggregates/MaxSizeForStatsAggregate.cpp

```diff
@@ -0,0 +1,269 @@
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    doUpdateSingleGroup(group, rows, args[0]);
+  }
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    doUpdate(groups, rows, args[0]);
+  }
+
+ protected:
+  void upsertOneAccumulator(
+      char* const group,
+      const BaseVector& inputVector,
+      std::vector<vector_size_t>& rowSizes,
+      vector_size_t idx) {
+    if (inputVector.isNullAt(idx)) {
+      return;
+    }
+
+    // Clear null
+    clearNull(group);
+    // Set max(current, this)
+    int64_t& cur = *value<int64_t>(group);
+    cur = std::max(cur, (int64_t)rowSizes[idx]);
+  }
+
+  void
+  doUpdate(char** groups, const SelectivityVector& rows, const VectorPtr& arg) {
+    DecodedVector decoded(*arg, rows, true);
+    auto indices = decoded.indices();
+    auto baseVector = decoded.base();
+
+    // TODO: Add test to verify that constant mapping encoding works w.r.t idx.
+    if (decoded.isConstantMapping() && decoded.isNullAt(0)) {
+      // nothing to do; all values are nulls
+      return;
+    }
+
+    auto numRows = arg->size();
+    std::vector<vector_size_t> elementSizes(numRows, 0);
+    std::vector<vector_size_t*> elementSizePtrs(numRows);
+    std::vector<IndexRange> elementIndices(numRows);
+
+    for (int i = 0; i < numRows; i++) {
+      elementIndices[i] = IndexRange{i, 1};
+      elementSizePtrs[i] = &elementSizes[i];
+    }
+
+    getVectorSerde()->estimateSerializedSize(
+        arg,
+        folly::Range(elementIndices.data(), numRows),
+        elementSizePtrs.data());
+
+    rows.applyToSelected([&](vector_size_t i) {
+      upsertOneAccumulator(groups[i], *baseVector, elementSizes, indices[i]);
+    });
+  }
+
+  void doUpdateSingleGroup(
+      char* group,
+      const SelectivityVector& rows,
+      const VectorPtr& arg) {
+    DecodedVector decoded(*arg, rows, true);
+    auto indices = decoded.indices();
+    auto baseVector = decoded.base();
+
+    auto numRows = baseVector->size();
+    std::vector<vector_size_t> elementSizes(numRows, 0);
+    std::vector<vector_size_t*> elementSizePtrs(numRows);
+    std::vector<IndexRange> elementIndices(numRows);
+
+    for (int i = 0; i < numRows; i++) {
+      elementIndices[i] = IndexRange{i, 1};
+      elementSizePtrs[i] = &elementSizes[i];
+    }
+
+    getVectorSerde()->estimateSerializedSize(
+        arg,
+        folly::Range(elementIndices.data(), numRows),
+        elementSizePtrs.data());
+
+    if (decoded.isConstantMapping()) {
+      if (decoded.isNullAt(0)) {
+        // nothing to do; all values are nulls
+        return;
+      }
+      upsertOneAccumulator(group, *baseVector, elementSizes, 0);
+      return;
+    }
+
+    rows.applyToSelected([&](vector_size_t i) {
+      upsertOneAccumulator(group, *baseVector, elementSizes, i);
```

Comment:
Would it be more efficient to calculate max over elementSizes then call upsertOneAccumulator once?

Reply:
Not quite because we still need to upsertOneAccumulator on each index "i" to ensure that we do not clear null.

So computing max first will not get around checking each index "i" for nullness.